database.cc

```diff
@@ -1847,25 +1827,34 @@ future<> database::do_apply(schema_ptr s, const frozen_mutation& m, tracing::tra
     auto uuid = m.column_family_id();
     auto& cf = find_column_family(uuid);
     if (!s->is_synced()) {
-        return make_exception_future<>(std::runtime_error(format("attempted to mutate using not synced schema of {}.{}, version={}",
-                s->ks_name(), s->cf_name(), s->version())));
+        throw std::runtime_error(format("attempted to mutate using not synced schema of {}.{}, version={}", s->ks_name(), s->cf_name(), s->version()));
     }
 
     sync = sync || db::commitlog::force_sync(s->wait_for_sync_to_commitlog());
 
     // Signal to view building code that a write is in progress,
     // so it knows when new writes start being sent to a new view.
     auto op = cf.write_in_progress();
-    if (cf.views().empty()) {
-        return apply_with_commitlog(std::move(s), cf, std::move(uuid), m, timeout, sync).finally([op = std::move(op)] { });
-    }
-    future<row_locker::lock_holder> f = cf.push_view_replica_updates(s, m, timeout, std::move(tr_state), get_reader_concurrency_semaphore());
-    return f.then([this, s = std::move(s), uuid = std::move(uuid), &m, timeout, &cf, op = std::move(op), sync] (row_locker::lock_holder lock) mutable {
-        return apply_with_commitlog(std::move(s), cf, std::move(uuid), m, timeout, sync).finally(
-                // Hold the local lock on the base-table partition or row
-                // taken before the read, until the update is done.
-                [lock = std::move(lock), op = std::move(op)] { });
-    });
+
+    row_locker::lock_holder lock;
+    if (!cf.views().empty()) {
+        lock = co_await cf.push_view_replica_updates(s, m, timeout, std::move(tr_state), get_reader_concurrency_semaphore());
+    }
+
+    // purposefully manually "inlined" apply_with_commitlog call here to reduce # coroutine
+    // frames.
```

Comment:
Does calling `apply_with_commitlog()` here add another allocation?

Reply:
Yes. Very carefully tweaked to minimise coroutine frame allocations.